src/vs/workbench/contrib/chat/browser/chatWidget.ts

```diff
@@ -1177,28 +1181,49 @@ export class ChatWidget extends Disposable implements IChatWidget {
 			const { promptInstructions } = this.inputPart.attachmentModel;
 			const instructionsEnabled = promptInstructions.featureEnabled;
 			if (instructionsEnabled) {
-				// instruction files may have nested child references to other prompt
-				// files that are resolved asynchronously, hence we need to wait for
-				// the entire prompt instruction tree to be processed
+				// prompt files may have nested child references to other prompt
+				// files that are resolved asynchronously, hence we need to wait
+				// for the entire prompt instruction tree to be processed
 				const instructionsStarted = performance.now();
 				await promptInstructions.allSettled();
 				// allow-any-unicode-next-line
 				this.logService.trace(`[⏱] instructions tree resolved in ${performance.now() - instructionsStarted}ms`);
 			}
 
-			let attachedContext = this.inputPart.getAttachedAndImplicitContext(this.viewModel.sessionId);
+			let attachedContext = await this.inputPart.getAttachedAndImplicitContext(this.viewModel.sessionId);
 			if (instructionsEnabled) {
 				const result = await this._handlePromptSlashCommand();
 				if (result.input !== undefined) {
 					input = result.input;
 				}
-				if (result.attachment) {
-					attachedContext.push(result.attachment);
+
+				const { attachment } = result;
+				if (attachment) {
+					attachedContext.push(attachment);
 				}
 
-				const promptFileChatVariable = attachedContext.filter(isPromptFileChatVariable);
-				if (promptFileChatVariable.length > 0) {
-					input = `Follow the prompt instructions from ${promptFileChatVariable.map(v => v.name).join(', ')}\n${input}`;
+				const promptFileVariables = attachedContext.filter(isPromptFileChatVariable);
+				if (promptFileVariables.length > 0) {
+					input = `Follow the prompt instructions from ${promptFileVariables.map(pick('name')).join(', ')}\n${input}`;
+
+					const allToolsMetadata = await this.getPromptFileToolsMetadata(promptFileVariables);
+
+					// if there are some tools defined in the prompt files, switch to
+					// the agent mode and select only the specified tools
+					if ((allToolsMetadata !== null) && (allToolsMetadata.length > 0)) {
+						const options: IToggleChatModeArgs = { mode: ChatMode.Agent };
+
+						// tools are currently only available in agent mode hence
+						// switch to the mode before updating the selected tools
+						await this.commandService.executeCommand(
```

Comment:
I think this won't handle the mode switching correctly when switching out of Edits mode, because we ask users to clear the session. If they answer no, then we should not submit the request. But you don't really have a good way to do this now because there's no way for you to know. Doesn't have to block the PR, can you open an issue for me to refactor?

Reply:
Sounds good will do.